samza-api/src/main/java/org/apache/samza/operators/windows/WindowInternal.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.samza.operators.windows;
+
+import org.apache.samza.annotation.InterfaceStability;
+import org.apache.samza.operators.data.MessageEnvelope;
+import org.apache.samza.operators.triggers.Trigger;
+
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+/**
+ *  Internal representation of a {@link Window}. This specifies default, early and late triggers for the {@link Window}
+ *  and whether to accumulate or discard previously emitted panes.
+ */
+@InterfaceStability.Unstable
+public final class WindowInternal<M extends MessageEnvelope, K, WV> implements Window<M, K, WindowKey<K>, WV, WindowPane<WindowKey<K>, WV>> {
+
+  public enum AccumulationMode { ACCUMULATING, DISCARDING }
+
+  private final Trigger defaultTrigger;
+
+  /*
+   * The function that is applied each time a {@link MessageEnvelope} is added to this window.
+   */
+  private final BiFunction<M, WV, WV> foldFunction;
+
+  /*
+   * The function that extracts the key from a {@link MessageEnvelope}
+   */
+  private final Function<M, K> keyExtractor;
+
+  /*
+   * The function that extracts the event time from a {@link MessageEnvelope}
+   */
+  private final Function<M, Long> eventTimeExtractor;
+
+  private Trigger earlyTrigger;
+
+  private Trigger lateTrigger;
+
+  private AccumulationMode mode;
+
+  WindowInternal(Trigger defaultTrigger, BiFunction<M, WV, WV> foldFunction, Function<M, K> keyExtractor, Function<M, Long> eventTimeExtractor) {
+    this.foldFunction = foldFunction;
+    this.eventTimeExtractor = eventTimeExtractor;
+    this.keyExtractor = keyExtractor;
+    this.defaultTrigger = defaultTrigger;
+  }
+
+  @Override
+  public Window<M, K, WindowKey<K>, WV, WindowPane<WindowKey<K>, WV>> setEarlyTrigger(Trigger trigger) {
```

Comment:
The new types read as follows:
```
public interface Window<M extends MessageEnvelope, K, WV, WM extends WindowPane<K, WV>> { }
```
where

WK is replaced with WindowKey<K>. That way we can remove one type parameter from the user API and simplify it.